eth2/types/src/eth_spec.rs

```diff
@@ -0,0 +1,91 @@
+// repository, updated to use explicit imports, and added additional values to the spec definition.
+use crate::beacon_state::BeaconState;
+use crate::newtypes::Epoch;
+use serde::{Deserialize, Serialize};
+use std::fmt::Debug;
+use typenum::{
+    Unsigned, U0, U1, U1024, U1099511627776, U128, U16, U16777216, U2048, U32, U4, U4096, U64,
+    U65536, U8, U8192,
+};
+
+pub trait EthSpec: 'static + Default + Sync + Send + Clone + Debug + PartialEq {
+    /*
+     * Unspecced values
+     */
+    type ShardCount: Unsigned + Clone + Sync + Send + Debug + PartialEq;
+    type MaxExecutionEnvironments: Unsigned + Clone + Sync + Send + Debug + PartialEq;
+
+    /*
+     * Constants
+     */
+    type JustificationBitsLength: Unsigned + Clone + Sync + Send + Debug + PartialEq + Default;
+    /*
+     * Misc
+     */
+    type MaxValidatorsPerCommittee: Unsigned + Clone + Sync + Send + Debug + PartialEq;
+    /*
+     * Initial values
+     */
+    type GenesisEpoch: Unsigned + Clone + Sync + Send + Debug + PartialEq;
+    /*
+     * Time parameters
+     */
+    type SlotsPerEpoch: Unsigned + Clone + Sync + Send + Debug + PartialEq;
+    type SlotsPerEth1VotingPeriod: Unsigned + Clone + Sync + Send + Debug + PartialEq;
+    type SlotsPerHistoricalRoot: Unsigned + Clone + Sync + Send + Debug + PartialEq;
+    /*
+     * State list lengths
+     */
+    type EpochsPerHistoricalVector: Unsigned + Clone + Sync + Send + Debug + PartialEq;
+    type EpochsPerSlashingsVector: Unsigned + Clone + Sync + Send + Debug + PartialEq;
+    type HistoricalRootsLimit: Unsigned + Clone + Sync + Send + Debug + PartialEq;
+    type ValidatorRegistryLimit: Unsigned + Clone + Sync + Send + Debug + PartialEq;
+    /*
+     * Max operations per block
+     */
+    type MaxProposerSlashings: Unsigned + Clone + Sync + Send + Debug + PartialEq;
+    type MaxAttesterSlashings: Unsigned + Clone + Sync + Send + Debug + PartialEq;
+    type MaxAttestations: Unsigned + Clone + Sync + Send + Debug + PartialEq;
+    type MaxDeposits: Unsigned + Clone + Sync + Send + Debug + PartialEq;
+    type MaxVoluntaryExits: Unsigned + Clone + Sync + Send + Debug + PartialEq;
+    /*
+     * Derived values (set these CAREFULLY)
+     */
+    /// The length of the `{previous,current}_epoch_attestations` lists.
+    ///
+    /// Must be set to `MaxAttestations * SlotsPerEpoch`
+    // NOTE: we could safely instantiate this by using type-level arithmetic, but doing
+    // so adds ~25s to the time required to type-check this crate
+    type MaxPendingAttestations: Unsigned + Clone + Sync + Send + Debug + PartialEq;
+}
+
+/// Ethereum Foundation specifications.
+///
+/// Spec v0.9.1
+#[derive(Clone, PartialEq, Debug, Default, Serialize, Deserialize)]
+pub struct MainnetEthSpec;
+
+impl EthSpec for MainnetEthSpec {
+    type ShardCount = U64;
+    type MaxExecutionEnvironments = U65536;
+    type JustificationBitsLength = U4;
+    type MaxValidatorsPerCommittee = U2048;
```

Comment:
Again, I'd pop out the ones we don't need for now and then you can add them slowly in. Also do we want a mainnetEthspec setup or the minimalistic one?

Reply:
If you're cool with it I'll just comment 'em.  I think either ethspec works good.

Reply:
yeah sounds good to me to just comment